dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1183 @@
+			}
+		})
+	}
+}
+
+func TestChainSync_SetPeerHead(t *testing.T) {
+	t.Parallel()
+
+	cs := newTestChainSync(t)
+
+	testPeer := peer.ID("noot")
+	hash := common.Hash{0xa, 0xb}
+	const number = 1000
+	ctrl := gomock.NewController(t)
+	mockBlockState := NewMockBlockState(ctrl)
+	mockHeader, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, 0,
+		types.NewDigest())
+	require.NoError(t, err)
+	mockBlockState.EXPECT().BestBlockHeader().Return(mockHeader, nil)
+	cs.blockState = mockBlockState
+
+	err = cs.setPeerHead(testPeer, hash, number)
+	require.NoError(t, err)
+
+	expected := &peerState{
+		who:    testPeer,
+		hash:   hash,
+		number: number,
+	}
+	require.Equal(t, expected, cs.peerState[testPeer])
+	require.Equal(t, expected, <-cs.workQueue)
+	require.True(t, cs.pendingBlocks.hasBlock(hash))
+
+	// test case where peer has a lower head than us, but they are on the same chain as us
+	cs.blockState = new(mocks.BlockState)
+	header, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, number,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*mocks.BlockState).On("BestBlockHeader").Return(header, nil)
+	fin, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, number-2,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*mocks.BlockState).On("GetHighestFinalisedHeader").Return(fin, nil)
+	cs.blockState.(*mocks.BlockState).On("GetHashByNumber", mock.AnythingOfType("uint")).Return(hash, nil)
+
+	err = cs.setPeerHead(testPeer, hash, number-1)
+	require.NoError(t, err)
+	expected = &peerState{
+		who:    testPeer,
+		hash:   hash,
+		number: number - 1,
+	}
+	require.Equal(t, expected, cs.peerState[testPeer])
+	select {
+	case <-cs.workQueue:
+		t.Fatal("should not put chain we already have into work queue")
+	default:
+	}
+
+	// test case where peer has a lower head than us, and they are on an invalid fork
+	cs.blockState = new(mocks.BlockState)
+	cs.blockState.(*mocks.BlockState).On("BestBlockHeader").Return(header, nil)
+	fin, err = types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, number,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*mocks.BlockState).On("GetHighestFinalisedHeader").Return(fin, nil)
+	cs.blockState.(*mocks.BlockState).On("GetHashByNumber", mock.AnythingOfType("uint")).Return(common.Hash{}, nil)
+
+	mockNetwork := NewMockNetwork(ctrl)
+	mockNetwork.EXPECT().ReportPeer(peerset.ReputationChange{
+		Value:  -4096,
+		Reason: "Bad block announcement",
+	}, peer.ID("noot"))
+	cs.network = mockNetwork
+	err = cs.setPeerHead(testPeer, hash, number-1)
+	require.True(t, errors.Is(err, errPeerOnInvalidFork))
+	expected = &peerState{
+		who:    testPeer,
+		hash:   hash,
+		number: number - 1,
+	}
+	require.Equal(t, expected, cs.peerState[testPeer])
+	select {
+	case <-cs.workQueue:
+		t.Fatal("should not put invalid fork into work queue")
+	default:
```

Comment:
same here. The production code might put it in the work queue but this won't trigger since the goroutine was slower than the test goroutine.